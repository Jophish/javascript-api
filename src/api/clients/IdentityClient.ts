export default class IdentityClient {
  
  save() {};
  update() {};
  delete() {};
}