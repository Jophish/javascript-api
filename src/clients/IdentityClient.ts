import Client from './Client';
import Identity, { IIdentity } from '../dtos/Identity';

export default class IdentityClient extends Client {
  /**
   * Saves an identity. It will return a promise resolving to the reason
   * if an error occurs, or an Identity otherwise.
   */
  save(identity: IIdentity): Promise<Identity> {
    try {
      const ident = new Identity(identity);
      return this.gateway.put(
        ident,
        `/trafficTypes/${ident.trafficTypeId}/environments/${ident.environmentId}/identities/${ident.key}`
      ).then((res) => {
        return new Identity(<any> res);
      });
    } catch (err) {
      return <any> this.gateway.rejectedReq(err);
    }
  };
  /**
   * Save a list of identities. It will return a promise resolving to the array
   * of the results of saving each identity. Each position will have an Identity
   * object, or an object with two properties if an error has occured: {
   *   data: { the identity object that caused the error },
   *   err: 'the error message'
   * }.
   * NOTE: Entities error results would be at the beginning of the resulting array
   */
  saveBulk(identities: Array<IIdentity>): Promise<Array<Identity | {
    data: string,
    err: Error
  }>> {
    let groups: {
      [groupName: string]: Array<IIdentity>
    } = {};
    let promises: Array<Promise<Identity>> = [];

    // Group the identities
    identities.forEach((i) => {
      let ident;
      let key;

      try {
        ident = new Identity(i);
        key = `${i.trafficTypeId}+${i.environmentId}`;
      } catch (err) {
        ident = {
          data: i,
          err
        };
      }

      if (key) { // If we have an Identity-like object
        if (!groups[key]) {
          groups[key] = [i];
        } else {
          groups[key].push(i);
        }
      } else { // If we don't have a the necessary data.
        promises.push(ident);
      }
    });
    // Save each group and keep promise reference.
    for (var key in groups) {
      let group = groups[key];

      const ttId = group[0].trafficTypeId;
      const envId = group[0].environmentId;

      promises.push(
        this.gateway.post(
          group,
          `/trafficTypes/${ttId}/environments/${envId}/identities`
        ).then((res: any) => {
          return res.objects.map(e => e = new Identity(e));
        })
      );
    }

    return Promise.all(promises);
  };
  /**
   * Update an identity. It will return a promise resolving to the reason
   * if an error occurs, or an Identity otherwise.
   */
  update(identity: IIdentity): Promise<Identity> {
    try {
      const ident = new Identity(identity);
      return this.gateway.patch(
        ident,
        `/trafficTypes/${ident.trafficTypeId}/environments/${ident.environmentId}/identities/${ident.key}`
      ).then((res) => {
        return new Identity(<any> res);
      });
    } catch (err) {
      return <any> this.gateway.rejectedReq(err);
    }
  };
  /**
   * Delete an identity. It will return a promise resolving to the reason
   * if an error occurs, or a boolean otherwise.
   */
  delete(identity: IIdentity): Promise<boolean> {
    try {
      const ident = new Identity(identity);
      return this.gateway.del(`/trafficTypes/${identity.trafficTypeId}/environments/${identity.environmentId}/identities/${identity.key}`);
    } catch (err) {
      return <any> this.gateway.rejectedReq(err);
    }
  };
}
