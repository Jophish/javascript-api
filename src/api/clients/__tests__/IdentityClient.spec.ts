jest.mock('../../http/gateway');

import IdentityClient from '../IdentityClient';

const NicoZelayaIdent = {
  'environmentId': 'STG',
  'key': 'NicoZelaya',
  'trafficTypeId': 'userTT',
  'values': {
    'username': 'NicoZelaya',
    'email': 'nico@split.io',
    'dob': '02/08/1989'
  }
};
const choculaIdent = {
  'environmentId': 'STG',
  'key': 'Chocula666',
  'trafficTypeId': 'machineTT',
  'values': {
    'username': 'Chocula666',
    'email': 'countchocula@vampires.ts',
    'dob': '24/07/1710'
  }
}

test('should be a class (function)', () => {
  expect(typeof IdentityClient).toBe('function');
});

test('instances should expose an API', () => {
  const identity = new IdentityClient();

  expect(identity.save).toBeDefined();
  expect(identity.saveBulk).toBeDefined();  
  expect(identity.update).toBeDefined();
  expect(identity.delete).toBeDefined();
});

test('it should be able to create an identity, returning a promise', () => {
  const identity = new IdentityClient();
  const createPromise = identity.save(NicoZelayaIdent);

  expect(createPromise.then).toBeDefined();
  expect(typeof createPromise.then).toBe('function');  

  return createPromise.then((res: any) => {
    expect(res.data).toMatchObject(NicoZelayaIdent);
    expect(res.response).toMatchSnapshot();
  });
});

test('it should be able to save a collection of identities, returning a promise', () => {
  const identity = new IdentityClient();
  const createBulkPromise = identity.saveBulk([
    NicoZelayaIdent, choculaIdent
  ]);

  expect(createBulkPromise.then).toBeDefined();
  expect(typeof createBulkPromise.then).toBe('function');  

  return createBulkPromise.then((res: any) => {
    // We should have two "groups" (combinations of TT & Env)
    expect(res.length).toBe(2);
    const nico = res[0];
    const chocula = res[1];

    // It should sent the data of the received identities
    expect(nico.data).toMatchObject([NicoZelayaIdent]);
    expect(chocula.data).toMatchObject([choculaIdent]);

    // And receive a response
    expect(nico.response.objects.length).toBe(1);
    expect(nico.response.failed.length).toBe(0);
    expect(nico.response.metadata).toBeDefined();
    expect(chocula.response.objects.length).toBe(1);
    expect(chocula.response.failed.length).toBe(0);
    expect(chocula.response.metadata).toBeDefined();

    // with those identities full data.
    expect(nico.response.objects[0]).toMatchObject(NicoZelayaIdent);
    expect(chocula.response.objects[0]).toMatchObject(choculaIdent);    
  });
});

test('it should be able to update an identity, returning a promise', () => {
  const identity = new IdentityClient();
  const nicoUpdate = Object.assign({
    email: 'nicolas.zelaya@split.io'
  }, NicoZelayaIdent);
  const updatePromise = identity.update(nicoUpdate);

  expect(updatePromise.then).toBeDefined();
  expect(typeof updatePromise.then).toBe('function');  

  return updatePromise.then((res: any) => {
    expect(res.data).toMatchObject(nicoUpdate);
    expect(res.response).toMatchSnapshot();
  });
});

test('it should be able to delete an identity, returning a promise', () => {
  const identity = new IdentityClient();
  const deletePromise = identity.delete(NicoZelayaIdent);

  expect(deletePromise.then).toBeDefined();
  expect(typeof deletePromise.then).toBe('function');

  return deletePromise.then((res) => {
    expect(res).toBe(true);
  });
});