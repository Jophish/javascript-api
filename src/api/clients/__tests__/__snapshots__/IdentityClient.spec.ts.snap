// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`it should be able to create an identity, returning a promise 1`] = `
Object {
  "environmentId": "STG",
  "key": "NicoZelaya",
  "organizationId": "SplitTesting",
  "timestamp": 1496102956729,
  "trafficTypeId": "userTT",
  "values": Object {
    "dob": "02/08/1989",
    "email": "nico@gmail.com",
    "username": "NicoZelaya",
  },
}
`;

exports[`it should be able to update an identity, returning a promise 1`] = `
Object {
  "environmentId": "STG",
  "key": "NicoZelaya",
  "organizationId": "SplitTesting",
  "timestamp": 1496102956729,
  "trafficTypeId": "userTT",
  "values": Object {
    "dob": "02/08/1989",
    "email": "nicolas.zelaya@gmail.com",
    "username": "NicoZelaya",
  },
}
`;
